feat(plugins/k8saudit): add header-based custom fields to events

sboschman opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature


<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area plugins

> /area registry

> /area build

> /area documentation

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:

For context see:
- [falco #2289](https://github.com/falcosecurity/falco/issues/2289)
- [falco #1704](https://github.com/falcosecurity/falco/issues/1704)
- [falco #2127](https://github.com/falcosecurity/falco/issues/2127)

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #206

**Special notes for your reviewer**:
Sorry for the spam, tried to open a 'draft pr', and hit the button to change pr type, ~but I guess that isn't supported by github on this repository~ nvm I am blind, found the sneaky little convert to draft link , The code needs a cleanup before a review. Just wanted a draft pr to get some feedback on the idea/concept and the plugin configuration.

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [03c2a91](https://github.com/falcosecurity/plugins/commits/03c2a91afbada0da1b94aae7eee7e13ec7a2f184) feat(plugins/k8saudit): add header-based custom fields to events

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Welcome @sboschman! It looks like this is your first PR to falcosecurity/plugins 🎉

poiana commented:
Hi @sboschman. Thanks for your PR.

I'm waiting for a [falcosecurity](https://github.com/orgs/falcosecurity/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Fplugins).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/plugins/pull/205#" title="Author self-approved">sboschman</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign mstemm for approval by writing `/assign @mstemm` in a comment. For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Fplugins).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/plugins/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["mstemm"]} -->

sboschman commented:
Hello @jasondellaluce and @leogr , as github automatically added you as reviewers, hopefully you can spare a moment to give some feedback on the idea/concept behind this PR. In the first place, is this feat a useful addition to the k8s-audit plugin at all, or do you see a better approach to add additional metadata to k8s audit events. And secondly, is the new `customFieldsHeaders` config option looking oke.

The actual code is not ready for a code review yet, but feel free to have a peek at it if that would help understanding the approach I am taking.

The readme snippet I added to express the idea behind the config option I added:
> The Kubernetes webhook audit backend configuration is basically a `kubeconfig`. This does not allow setting custom HTTP headers to sent along with the POST requests. However, it does allow defining authentication by username+password or certificate.
A reverse proxy in front of the plugin embedded webserver for authentication allows for passing authentication info as 'forwarded' headers. Based on the headers passed by a reverse proxy this plugin can add custom fields to each event. For example the id/name of the originating Kubernetes cluster of the event, as the Kubernetes audit event itself does not contain cluster identification.
The custom fields are not part of the defined supported extraction fields. To use custom fields in rules a json reference has to be used, e.g. `%json.value[/cluster_name]`.

The actual config yaml:

```yaml
plugins:
  - name: k8saudit
    library_path: libk8saudit.so
    init_config:
      sslCertificate: /etc/falco/falco.pem
      customFieldsHeaders:
        # Example HTTP header sent with the request:
        # X-WebAuth-User: cheese-cluster
        - header: "X-WebAuth-User"
          customFields:
            - cluster_name
        # Example HTTP header sent with the request:\
        # X-Forwarded-TlS-Client-Cert-Info: Subject=\"DC=org,DC=cheese,OU=CheeseFactory,OU=Packaging,CN=*.example.com\"
        - header: "X-Forwarded-TlS-Client-Cert-Info"
          pattern: "Subject=\"DC=[A-Za-z0-9]+,DC=[A-Za-z0-9]+,OU=([A-Za-z0-9]+),OU=([A-Za-z0-9]+),.+\""
          customFields:
            - division
            - team
```
  
Another added snippet from the readme to describe the new config options:
>**Initialization Config**:
>- `customFieldsHeaders`: List of custom fields to add to every audit event based on HTTP request headers (Optional)

>**HTTP Header-based custom fields Config**
>- `header`: HTTP header name to extract custom field(s) from (Required)
>- `pattern`: Regex pattern containing capture groups, in order each capture group corresponds to a custom field (Optional)
>- `customFields`: List of custom fields to add, without `pattern` only a single custom field can be defined (Required)

poiana commented:
Issues go stale after 90d of inactivity.

Mark the issue as fresh with `/remove-lifecycle stale`.

Stale issues rot after an additional 30d of inactivity and eventually close.

If this issue is safe to close now please do so with `/close`.

Provide feedback via https://github.com/falcosecurity/community.

/lifecycle stale

sboschman commented:
/remove-lifecycle stale

Just waiting for feedback on the idea of adding custom fields based on headers and if such a feature would be accepted, before spending more time on the code.

This feature should be seen as part of a larger effort to build a centralised Kubernetes audit log system with Falco. In which multiple k8s clusters sent their audit logs to this central system for analysis and further processing (with the Falco stack). In such a setup we need a way to enrich the incoming k8s audit logs with additional metadata not present in the audit log messages themselves. E.g. the name of the originating k8s cluster.
